File: Leetcode 3001-3100/3031/3031.java
// Leetcode 3031: Minimum Time to Revert Word to InitialState II
// https://leetcode.com/problems/minimum-time-to-revert-word-to-initial-state-ii/
// Solved on 12th of December, 2025
class Solution {
    /**
     * Calculates the minimum time to revert a word to its initial state.
     * @param word The input word.
     * @param k The number of characters to remove from the beginning in each step.
     * @return The minimum number of steps required.
     */
    public int minimumTimeToInitialState(String word, int k) {
        int n = word.length();
        char[] s = word.toCharArray();
        int[] z = new int[n];
        int left = 0;
        int right = 0;

        for (int i = 1; i < n; i++) {
            if (i <= right) {
                z[i] = Math.min(right - i + 1, z[i - left]);
            }
            while (i + z[i] < n && s[z[i]] == s[i + z[i]]) {
                z[i]++;
            }
            if (i + z[i] - 1 > right) {
                left = i;
                right = i + z[i] - 1;
            }
        }

        for (int i = k; i < n; i += k) {
            if (z[i] == n - i) {
                return i / k;
            }
        }

        return (n + k - 1) / k;
    }
}

File: Leetcode 3001-3100/3031/3031.py
# Leetcode 3031: Minimum Time to Revert Word to Initial State II
# https://leetcode.com/problems/minimum-time-to-revert-word-to-initial-state-ii/
# Solved on 12th of December, 2025
class Solution:
    def minimumTimetoInitialState(self, word: str, k: int) -> int:
        """
        Calculates the minimum time to revert a word to its initial state.

        Args:
            word (str): The input word.
            k (int): The number of characters to remove from the beginning in each operation.

        Returns:
            int: The minimum number of operations required to revert the word to its initial state.
        """
        n = len(word)
        lps = [0] * n
        currentLength = 0
        i = 1

        while i < n:
            if word[i] == word[currentLength]:
                currentLength += 1
                lps[i] = currentLength
                i += 1
            else:
                if currentLength != 0:
                    currentLength = lps[currentLength - 1]
                else:
                    lps[i] = 0
                    i += 1

        maxLength = lps[n - 1]

        while maxLength > 0:
            if (n - maxLength) % k == 0:
                return (n - maxLength) // k
            maxLength = lps[maxLength - 1]

        return (n + k - 1) // k